File: challenge10/src/main/kotlin/BoardLocation.kt
class BoardLocation(private val locations: List<Location>) {

    private var locationIndex = 0

    var hasPassedGo = false
        private set

    val currentLocation get() = locations[locationIndex]

    constructor (_locations:List<Location>, _locationIndex:Int):this(_locations) {

        if (_locationIndex > locations.lastIndex) {
            this.locationIndex = _locationIndex % locations.size
            this.hasPassedGo = true
        } else {
            this.locationIndex = _locationIndex
        }

    }

    operator fun plus(dice:Dice):BoardLocation {
        val newLocationIndex = (this.locationIndex + dice.totalValue)
        return BoardLocation(this.locations, newLocationIndex)
    }
}

File: challenge10swift/challenge10/challenge10/BoardLocation.swift
//
//  BoardLocation.swift
//  challenge10
//
//  Created by Michael Neilens on 15/08/2019.
//  Copyright © 2019 Michael Neilens. All rights reserved.
//

struct BoardLocation {
    
    private let locationIndex:Int
    private let locations:Array<Location>
    
    let hasPassedGo:Bool
    var currentLocation:Location { return locations[locationIndex] }
    
    init(locations _locations:Array<Location>) {
        self.locations = _locations
        self.locationIndex = 0
        self.hasPassedGo = false
    }
    
    init(locations _locations:Array<Location>, locationIndex _locationIndex:Int) {
        self.locations = _locations
        if _locationIndex > (locations.count - 1 ) {
            self.locationIndex = _locationIndex % locations.count
            self.hasPassedGo = true
        } else {
            self.locationIndex = _locationIndex
            self.hasPassedGo = false
        }
    }
    
    func move(using dice:Dice) -> BoardLocation {
        let newLocationIndex = (locationIndex + dice.total)
        return BoardLocation(locations:locations, locationIndex:newLocationIndex)
    }
}
